src/sparse_merkle_tree/smt.rs

```diff
@@ -1362,24 +1364,31 @@ impl SparseMerkleTree {
 
     pub fn prepare_queries_with_proof_map(
         proof: &Proof,
-    ) -> HashMap<Vec<bool>, QueryProofWithProof> {
+    ) -> Result<HashMap<Vec<bool>, QueryProofWithProof>, SMTError> {
         let mut queries_with_proof: HashMap<Vec<bool>, QueryProofWithProof> = HashMap::new();
-        for query in &proof.queries {
-            let binary_bitmap = utils::strip_left_false(&utils::bytes_to_bools(&query.bitmap));
-            let binary_path = utils::bytes_to_bools(&query.pair.0)[..binary_bitmap.len()].to_vec();
-
-            queries_with_proof.insert(
-                binary_path,
-                QueryProofWithProof::new_with_pair(
-                    Arc::clone(&query.pair),
-                    &binary_bitmap,
-                    &[],
-                    &[],
-                ),
-            );
+        if !proof.queries.is_empty() {
+            let max_binary_bitmap_len = proof.queries[0].key().len() * 8; // converts key_length to bits
+            for query in &proof.queries {
+                let binary_bitmap = utils::strip_left_false(&utils::bytes_to_bools(&query.bitmap));
+                if binary_bitmap.len() > max_binary_bitmap_len {
+                    return Err(SMTError::InvalidBitmapLen);
+                }
+                let binary_path =
+                    utils::bytes_to_bools(query.key())[..binary_bitmap.len()].to_vec();
```

Comment:
wouldn't it be sufficient to just compare length of `utils::bytes_to_bools(query.key())` and `binary_bitmap.len()`?

Reply:
Thank you for that, it is better and more clear. Done.